docs/hypervisors/aws.md

```diff
@@ -0,0 +1,32 @@
+# Amazon Web Services - Elastic Compute Cloud (EC2)
+
+EC2 is a "web service that provides resizable compute capacity in the cloud."
+
+[EC2 site](https://aws.amazon.com/ec2/).
+
+# AWS Keys
+
+For any particular Beaker run, a new AWS key with a name of the form
+
+    Beaker-<username>-<sanitized_hostname>-<aws_keyname_modifier>-<readable_timestamp>
+
+will be created at the beginning of the run, & cleaned up at the end of the run.
+
+Everything up to _aws_keyname_modifier_ will be the same if run from the same
+user on the same machine no matter when it's run.
+
+_readable_timestamp_'s most fine grained unit is seconds. This means that if you
+have two Beaker runs kick off in the same second on the same host (which can
+happen often in CI environments), you can get collisions.  This will show
+itself through the `AWS::EC2::Errors::InvalidKeyPair::Duplicate` exception.
+
+The way to deal with this is to use the _aws_keyname_modifier_ variable.
+
+You can set the _aws_keyname_modifier_ variable in the host file in the options
+section:
+
+    CONFIG:
+      aws_keyname_modifier => unique_value,
+
+In CI systems, setting this to something like job name would make it unique enough
```

Comment:
Is JOB_NAME, what is recommended here? I'm a bit wary, so I went and tested what that string results to in our tests. Here is a JOB_NAME from one of our master runs:

```
+ echo 'Testing JOB_NAME'
Testing JOB_NAME
+ echo enterprise_puppetdb_integration-system-puppetdb_full-master/INSTALL_TYPE=install,LAYOUT=ec2-west-el7-64mda-el7-64a,PUPPETDB_DATABASE=embedded_db,label=beaker
enterprise_puppetdb_integration-system-puppetdb_full-master/INSTALL_TYPE=install,LAYOUT=ec2-west-el7-64mda-el7-64a,PUPPETDB_DATABASE=embedded_db,label=beaker
```

I'm not sure from memory what the rules are but we need to make sure this is going to be a valid resultant string that AWS would accept as a keypair name. I notice already we are being cautious around hostnames, and replacing the dots with hyphens from the hostname in #key_name_prefix. Those strings can sometimes contain weird stuff also.

Something to notice here, is that there is no other uniqueness element in JOB_NAME, the job number doesn't get included for example. If two jobs are running in parallel, they can still collide if JOB_NAME is the only thing used, and the tests start in the same second.